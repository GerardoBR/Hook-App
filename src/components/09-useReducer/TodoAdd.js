import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ {description }, handleInputChange , reset] =useForm({
        description: ''
    });
    const handleAddSumbit =(e)=>{
        e.preventDefault();
        if(description.trim().length<=1){
            return ;
        }
        const newTodo = {
                id : new Date().getTime(),
                desc : description,
                done : false
        }
        // const actionAddTodo ={
        //     type : 'add',
        //     payload : newTodo
        // }
        handleAddTodo(newTodo);
        reset();
    }
    return (
        <>
            <h4>Agregar To do</h4>
                    <hr/>

                    <form onSubmit = { handleAddSumbit }>
                        <input 
                            type="text"  
                            className = " form-control "
                            name = "description" 
                            placeholder = "tarea a hacer" 
                            value = { description }
                            onChange = { handleInputChange }
                        />
                        <button 
                            type ="submit"
                            className = "  form-control btn btn-outline-primary mt-1 btn-block"
                        > 
                            Agregar
                        </button>
                    </form>
        </>
    )
}
