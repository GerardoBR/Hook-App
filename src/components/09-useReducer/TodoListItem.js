import React from 'react'

export const TodoListItem = ( {todo , hanldeToggle , handleDelete ,index} ) => {
    return (
        <li 
            key ={ todo.id }
            className="list-group-item"
        > 
        <p 
            className={ `${todo.done  && 'complete'}` }
            onClick = { ()=>hanldeToggle(todo.id)}
        
        >{ index }. { todo.desc }</p>
            <button className= "btn btn-danger" onClick= { ()=>{ handleDelete(todo.id)  } }> Borrar</button>
        </li>
    )
}
