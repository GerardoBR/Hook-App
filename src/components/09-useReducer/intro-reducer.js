// console.log("hola mundo")
const initialSatet =[{
    id : 1,
    todo : 'comprar jamon',
    done : false
}];


const todoReducer =(state = initialSatet, action)=>{

    if ( action?.type === 'agregar'){
        return [...state,action.payload];
    }
    return state;
}
 let todos = todoReducer();

//  AÃ±adir un nuevo elemento al todo
// No usar push en react 
const newTodo ={
    id : 1,
    todo : 'comprar jugo',
    done : false
};

const addToDoAction ={
    type : 'agregar',
    payload : newTodo
}
todos = todoReducer( todos , addToDoAction);

 console.log(todos);
 