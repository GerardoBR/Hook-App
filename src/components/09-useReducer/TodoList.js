import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ( {todos ,hanldeToggle , handleDelete} ) => {
    return (
        <div>
            <ul className="list-group">
                {
                    
                    todos.map( ( todo, i ) =>(
                      <TodoListItem
                        todo = { todo }
                        handleDelete = { handleDelete }
                        hanldeToggle = { hanldeToggle }
                        index = { i+1 }
                        key ={ todo.id }

                      />
                        
                    ))
                }
            </ul> 
        </div>
    )
}
