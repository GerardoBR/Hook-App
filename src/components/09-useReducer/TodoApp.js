import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './toReducer';
import './todo.css';
// import { useForm } from '../../hooks/useForm';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';


const init =()=>{
    // return [{
    //     id : new Date().getTime(),
    //     desc : ' aprender React',
    //     done : false
    // }];
    return JSON.parse(localStorage.getItem('todos')) || [];

}
export const TodoApp = () => {
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    

    // console.log(description);
    useEffect(()=>{
        localStorage.setItem('todos' , JSON.stringify(todos));
    },[todos]);

    const handleDelete = (todoid)=>{
        const actionDelete={
            type: 'delete',
            payload : todoid
        }
        dispatch( actionDelete );
    }
    const hanldeToggle=(todoid)=>{

        dispatch({
            type: 'toggle',
            payload : todoid
        })
    }

    // const handleAddSumbit =(e)=>{
    //     e.preventDefault();
        
    //     const newTodo = {
    //         id : new Date().getTime(),
    //         desc : description,
    //         done : false
    //     }
    //     const actionAddTodo ={
    //         type : 'add',
    //         payload : newTodo
    //     }
    //     dispatch(actionAddTodo);
    //     reset();
    // }
    const handleAddTodo  = ( newTodo)=>{

        dispatch({
            type : 'add',
            payload : newTodo
        })
    }

    return (
        <div>
            <h1>TodoApp : ( { todos.length } )</h1>
            <hr/>
            <div className=" row">
                <div className = "col-6">
                    <TodoList 
                        todos = { todos }
                        hanldeToggle = { hanldeToggle }
                        handleDelete = { handleDelete }
                       
                    
                    />
                </div>
                <div className="col-6">
                    <TodoAdd 
                        handleAddTodo  = { handleAddTodo }
                    />
                </div>
                
            </div>
        </div>
    )
}


