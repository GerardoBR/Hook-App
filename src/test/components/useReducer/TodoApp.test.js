import { mount, shallow } from "enzyme";
import { TodoApp } from "../../../components/09-useReducer/TodoApp";
import { demoTodo } from "../../fixture/demoTodo";
import { act } from '@testing-library/react'
describe(' Pruebas sobre <TodoApp/>', () => {

    const wrapper = shallow(
        <TodoApp
        />
    )
    Storage.prototype.setItem= jest.fn(()=>{});

    test('debe de hacer Matchsnapshoot', () => {
        expect ( wrapper ).toMatchSnapshot();
    });

    test('debe de agregar un TODO', () => {
        const wrapper  = mount(<TodoApp/>);
        act(()=>{
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[1]);
        })

        expect ( wrapper.find('h1').text().trim()).toBe('TodoApp : ( 2 )');
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);
        // expect( localStorage.setItem ).toHaveBeenCalledWith({});
    })
    
    test('debe de eliminar un TODO', () => {
        
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[0]);
        expect ( wrapper.find('h1').text().trim()).toBe('TodoApp : ( 1 )');

        wrapper.find('TodoList').prop('handleDelete')(demoTodo[0].id);
        expect ( wrapper.find('h1').text().trim()).toBe('TodoApp : ( 0 )');

    })
    
    
});

