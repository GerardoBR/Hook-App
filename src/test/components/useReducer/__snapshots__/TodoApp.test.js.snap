// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Pruebas sobre <TodoApp/> debe de hacer Matchsnapshoot 1`] = `
<div>
  <h1>
    TodoApp : ( 
    0
     )
  </h1>
  <hr />
  <div
    className=" row"
  >
    <div
      className="col-6"
    >
      <TodoList
        handleDelete={[Function]}
        hanldeToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-6"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
